### DISCUSSION TITLE
Add isValid() to input widgets?

### DISCUSSION THREAD

**Discussion by ukoehler** [zauberzeug/nicegui#924] [Ideas / Feature Requests]
Hi,

I' m new to NiceGui and the underlying technology. So maybe I just do not know how to do this. When creating a ui.number widget with min = 0 and max = 100, the validation can be set accordingly:

```py
def parameterChange():
    print("parameterChange")
    if paramOne and paramOne.value >= 0 and paramOne.value <= 100:
        parameterLabel.set_text(f"You chose parameter 1: {int(paramOne.value)}")
        print(f"You chose parameter 1: {int(paramOne.value)}")

if __name__ in {"__main__", "__mp_main__"}:
    ui.label("Parameter 1 blih blah blub")
    paramOne = ui.number(label='Parameter 1', format='%i', value=50, min=0, max=100, step=1,
        on_change=lambda e: parameterChange(),
        validation={'Input too large': lambda value: value <= 100,
                    'Input too small': lambda value: value >= 0} )
    parameterLabel = ui.label("Choosen parameters")
    parameterChange()
    ui.run()
```
However, I can still enter 101 directly. I catch that case in parameterChange(), but I have no access to paramOne.min or paramOne.max nor can I find a way to check if the input is valid. The above code does work, however, I have to duplicate magic numbers. A function like paramOne.isValid() would be the best solution in my eyes.

Cheers

Uwe

**Comment by falkoschindler** [comment]
Yes, an `is_valid` method or property would be helpful. Some thoughts about it:

- It should evaluate the combination of `validation` and `min`/`max` parameters:
    ```py
    ui.number(min=0, max=100, value=101)  # invalid
    ui.number(min=0, max=100, value=50, validation={'Too small!': lambda value: value > 80})  # invalid
    ```
- The property should be bindable.
- We should add `is_valid` to _all_ elements with a `validation` parameter: `ui.number`, `ui.input` and `ui.textarea`.
- A `Validatable` mixin might be a good way to avoid lots of code duplication.

Would you like to open a pull request for it? You can start with a small proof-of-concept and we improve on it iteratively until it's ready to be merged.

**Comment by ukoehler** [comment]
Hi @falkoschindler .

I am flattered by your trust, but I could only achieve that in C++. I am still in the process of learning Python and I am barely able to use NiceGui, but not able to extend it at the moment. I can provide suggestions based on my experience with C++ based GUI frameworks, though.

**Comment by falkoschindler** [comment]
I just reviewed https://github.com/zauberzeug/nicegui/pull/894, which is closely related to this feature request.

**Reply by falkoschindler** [reply]
Now I pushed PR #924, covering the number limits that haven't been part of #894.

**Comment by falkoschindler** [comment]
This feature is implemented via PR #894 and PR #924 and will be released in 1.2.22.
